examples/reference/custom_components/ReactComponent.md

```diff
@@ -0,0 +1,460 @@
+# `ReactComponent`
+
+`ReactComponent` simplifies the creation of custom Panel components using [React](https://react.dev/).
+
+```pyodide
+import panel as pn
```

Comment:
I would like to simplify the example and just use `state` inside the `render` function. But it seems not possible: `TypeError: Cannot read properties of null (reading 'useState')`.

I would recommend it becomes possible.

![image](https://github.com/holoviz/panel/assets/42288570/775cde9e-636f-49f5-b1f2-e6dad6d279e2)




```python
import panel as pn
import param

from panel.custom import ReactComponent

pn.extension()

class CounterButton(ReactComponent):

    value = param.Integer()

    _esm = """
    export function render({state}) {
        const [value, setValue] = state.value;
        return (
            <button onClick={e => setValue(value+1)}>
            count is {value}
            </button>
        );
    }
    """

CounterButton().servable()
```

Reply:
Could fix this one by treating `render` as if it actually were a React component. Don't think there's any downside. This also automatically adds support for mount and unmount lifecycle handlers because you can do:

```javascript
    export function render({state}) {
        React.useEffect(() => {
          console.log('Mounted')
          return () => { console.log('Unmounted') }
        }, [])
        const [value, setValue] = state.value;
        return (
            <button onClick={e => setValue(value+1)}>
            count is {value}
            </button>
        );
    }
```

Reply:
Done.